Font and draw_text adjustments

WaywardHeart opened the pull request:
Brute forcing the font size calculation to get the proper dpi, although the kerning/letter spacing still isn't identical to RPGM.

Removing the leftmost column of pixels from draw_text calls is weird, but it's what RPGM does and wasn't hard to implement.

~~Some fonts that look smooth in RPGM look kind of ugly in mkxp-z. I think upgrading SDL_ttf to at least 2.20.0 for TTF_RenderUTF8_LCD would probably fix that.~~

WaywardHeart committed:
Don't let drawn text move past the top of the rect

WaywardHeart committed:
Limit squeezing text to a minimum of 50%

Not quite accurate to what VX Ace is doing, but I think that's mostly because of kerning differences.

WaywardHeart commented:
It looks like force pushing breaks the autobuild, and I can't restart it without making a new PR. I'll make sure to avoid doing that in the future. It just seemed like a waste to tack on an extra commit to clean up a leftover from testing when the PR was so fresh.

WaywardHeart commented:
Never mind the TTF_RenderUTF8_LCD, and on the text quality. It didn't change how the text looked, and the background alpha also affects the text. The text quality problem was from upscaling the game.

Edit: Checked out the scaling options and smoothScaling fixes it. I'm not sure why I didn't have it on in the first place.

Splendide-Imaginarius commented:
> It looks like force pushing breaks the autobuild, and I can't restart it without making a new PR.

Force-pushing is fine, and is actually preferred if it makes the Git history cleaner. I think what you hit is a GitHub Actions abuse mitigation where sometimes it makes a maintainer approve CI runs before they start.

Splendide-Imaginarius commented:
> Checked out the scaling options and smoothScaling fixes it. I'm not sure why I didn't have it on in the first place.

Lanczos might actually look better.

Eblo commented:
Regarding 55d536785a9bdb43fd25da315b8d906f98b58c0f, I have run into this myself before. I unknowingly drew excessively long strings to an invisible HUD element. This indeed resulted in texture corruptions in seemingly random unrelated areas.

Eblo commented:
> Some fonts that look smooth in RPGM look kind of ugly in mkxp-z. I think upgrading SDL_ttf to at least 2.20.0 for TTF_RenderUTF8_LCD would probably fix that.

I tried looking into this. I can verify that it does yield LCD subpixel style results, but it insists on having a background color. I've played around but to no avail. There is no function I can see that doesn't take in a background color and use it. [This discourse post from earlier this year](https://discourse.libsdl.org/t/ttf-renderutf8-lcd-with-no-background/43081) echoes the same issue, but there are no responses.

![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/80be085b-a755-4260-b58a-5d6012c585d0)

Background aside, I'm still not sure if it really fixes the ugliness of some fonts that look fine in RPG Maker. It doesn't look like RPG Maker font rendering uses LCD rendering; in fact, sometimes the glyphs are simply shaped differently between that implementation and mkxp.

WaywardHeart committed:
Use the first version of the font we find

WaywardHeart commented:
Yeah, I put in a later comment that the ugliness was actually me enlarging the game and not having smoothScaling on. I also just pushed something that might fix some differently shaped glyphs.

So I took another look at border opacity, and making it faithfully imitate RPG Maker is proving challenging. RPG Maker's comes out paler than my attempts for some reason. It's especially apparent with transparent text because the "outline" is supposed to show through the text in that case.

Also, I'm thinking about ripping out the actual drawing part of the function, adding a constructor to turn the surface into a bitmap, and just letting stretchBlt handle it. Any thoughts on that?

Splendide-Imaginarius commented:
> Also, I'm thinking about ripping out the actual drawing part of the function, adding a constructor to turn the surface into a bitmap, and just letting stretchBlt handle it. Any thoughts on that?

@WaywardHeart I would very much be in favor of having a constructor to create a Bitmap from an SDL software surface; it would enable some additional very useful functionality (e.g. the ability to create a software surface on a background thread, and then convert it to a Bitmap later on the main thread).

I'm not sure about the relative tradeoffs in using `stretchBlt` for this, can you elaborate what the benefits and drawbacks would be?

WaywardHeart commented:
I'd say the first benefit is that I just found out that the drawText code checked if it was drawing on a tainted rect, and the from surface part of stretchBlt isn't... Yeah, I'm just going to consolidate the code. That never would have happened if it was all in one place.

WaywardHeart committed:
Test for an invalid rect in drawText

Splendide-Imaginarius commented:
Could we have a `customScript` test for this PR too? This seems especially prone to regressions.

Splendide-Imaginarius commented:
> @WaywardHeart I would very much be in favor of having a constructor to create a Bitmap from an SDL software surface; it would enable some additional very useful functionality (e.g. the ability to create a software surface on a background thread, and then convert it to a Bitmap later on the main thread).

Implemented in https://github.com/mkxp-z/mkxp-z/pull/139, does that meet your needs for this?

WaywardHeart committed:
Match some outline positioning behavior from RGSS and enable transparent outlines

Outlines should 1) Not change the position of the text and 2) Be visible even if the text is cut off from reaching the edge of the rect, without leaving the rect itself.
We're currently not drawing full text for the outlines like RGSS does, so we don't perfectly succeed at (2), but if we ever start doing that then the logic is already there.

Text transparency with outlines present still doesn't match Enterbrain's implementation, but I gave up on trying to figure that out.

WaywardHeart committed:
Only draw enough of the text to fill the rect we were given

WaywardHeart commented:
Rebased onto #98, partially because there are code conflicts otherwise and partially because this is a big use case for casually testing it.

I gave up on replicating the quirks in opacity and squeezing, so you get this basic version with some comments about how they're supposed to work. Although, I've been doing all of the testing in macOS and recently noticed some color inaccuracies that aren't present in an older build I have lying around, so maybe there's just something wrong with my build of SDL and the opacity thing actually isn't complicated?

Text is sometimes a pixel lower than it should be. This is caused by a rounding difference in where freetype places the baseline, so there's nothing I can do about it without digging into there.


> > @WaywardHeart I would very much be in favor of having a constructor to create a Bitmap from an SDL software surface; it would enable some additional very useful functionality (e.g. the ability to create a software surface on a background thread, and then convert it to a Bitmap later on the main thread).
> 
> Implemented in #139, does that meet your needs for this?

Not really, but I fixed it. I do need it to be able to handle mega surfaces, and having an option to keep it as a mega surface is a lot more useful than having it not assume ownership of the surface.

Eblo commented:
Here are some comparisons with these fixes (whole-string draw on top, char-by-char on bottom):

Vanilla RPG Maker VX Ace
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/16665bc4-e352-46ac-be85-577fb29c60b5)

mkxp-z normally
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/829aa4c3-f28d-4890-a9cf-2add5be546c6)

This fix
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/d7856234-3f75-4121-a56b-71d9ac24ff2a)

mkxp-z compared to vanilla
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/4d45c113-96d9-4683-8478-2716ef8c3750)

This fix compared to vanilla
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/d4c0761c-7929-48a7-b955-6a24a5b7f678)

This fix compared to mkxp-z
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/bfb8c53f-ee09-4d68-b2ed-b21304d9a811)


While this is not exact (I'm sure getting it all the way there would cause even more pain), this is far far *far* closer to the mark.

Splendide-Imaginarius commented:
> Here are some comparisons with these fixes (whole-string draw on top, char-by-char on bottom):

@Eblo can you post the code necessary to reproduce those images?

Eblo commented:
> > Here are some comparisons with these fixes (whole-string draw on top, char-by-char on bottom):
> 
> @Eblo can you post the code necessary to reproduce those images?

Sure. Note that the comparison images were just layered in image editing software and were not actually rendered like that, as a single executable wouldn't be able to draw with and without the adjustments. I overwrote this method in the vanilla VXA `Scene_Title` class:
``` 
  #--------------------------------------------------------------------------
  # * Create Background
  #--------------------------------------------------------------------------
  def create_background
    @sprite1 = Sprite.new
    @sprite1.bitmap = Bitmap.new(Graphics.width, Graphics.height)
    c = 128
    @background = Plane.new
    w = h = 16
    @background.bitmap = Bitmap.new(w, h)
    @background.bitmap.fill_rect(@background.bitmap.rect, Color.new(c, c, c))
    r = Color.new(255, 0, 0)
    @background.bitmap.fill_rect(0, 0, w, 1, r)
    @background.bitmap.fill_rect(0, h - 1, w, 1, r)
    @background.bitmap.fill_rect(0, 0, 1, h, r)
    @background.bitmap.fill_rect(w - 1, 0, 1, h, r)
    @background.z = @sprite1.z - 1
    s = "The quick brown fox jumps over the lazy dog"
    pos = Rect.new(0, 0, @sprite1.width, Graphics.height)
    @sprite1.bitmap.draw_text(pos, s)
    pos.y += 20
    s.each_char do |char|
      size = @sprite1.bitmap.text_size(char)
      @sprite1.bitmap.draw_text(pos, char)
      pos.x += size.width
    end
  end
```

Splendide-Imaginarius commented:
Thanks @Eblo. What's the license on that snippet? Is it something you wrote, or a derivative of Enterbrain code?

WaywardHeart committed:
Hack around an inconsistency in the width returned by Bitmap.text_size

freetype sometimes treats the last character of the string as being a pixel wider than it should be. Adding a space at the end and then removing it's width should make character-by-character text more accurate.

WaywardHeart committed:
Add config options to control how the font size is determined and toggle font kerning.

Some Pokemon Essentials games were explicitly made for mkxp. Text rendered using the old method tends to be smaller than what the new one produces, which means text in those games tends to get cut off with it.

Since Pokemon Essentials uses RGSS1, the setting defaults to the old method for RGSS1, and the new method for RGSS2 and RGSS3.

WaywardHeart commented:
I just downloaded Pokemon Flux to check out a couple things, and the font size change definitely needs to be gated behind a config option. I'm defaulting it to the old method for XP games, since that's what Pokemon Essentials uses.

Also, I noticed an inconsistency in the bitmap.text_size function. The last character is sometimes treated as being a pixel wider than it should be. Working around that will probably make char-by-char text more accurate.

Eblo commented:
> Thanks @Eblo. What's the license on that snippet? Is it something you wrote, or a derivative of Enterbrain code?

It's just a quickie I slapped together. We can call it unlicensed or WTFPL, or whatever is more convenient to pass around. Any RGSS-experienced user could probably have come up with something equivalent by looking at the images and thinking it through, so I'm not worried about that snippet being "stolen".

Splendide-Imaginarius commented:
@Eblo Is licensing that snippet as CC0 okay?

Eblo commented:
> @Eblo Is licensing that snippet as CC0 okay?

That's fine by me.

Splendide-Imaginarius commented:
I just gave this PR a try, with the prologue of the first game I'm remastering. In terms of spacing accuracy, I am blown away, this is a near-perfect match for Enterbrain's runtime. I genuinely don't think I'd be able to perceive whatever difference remains without zooming in and looking pixel by pixel. Bravo!

However, I am seeing major graphical glitches in text rendering when Hires mode is enabled. @WaywardHeart have you tested this PR with Hires mode? I can try to post a test case if needed, but I'm hoping the glitch will be obvious enough that you can fix it without needing help reproducing it.

WaywardHeart commented:
I had an immediate suspicion what you were talking about, and it's what I thought it was. `drawText` is one of the functions where you decided to operate on both bitmaps, and I didn't change that or temporarily remove the hires version before blitting. I think I planned on asking you and forgot about it. I just added a commit to return after calling the hires function, let me know if drawing to both was an intentional design decision and I can do the other one.

Splendide-Imaginarius commented:
Thanks @WaywardHeart, yeah that fixed it.

So, this PR works very well in testing now. Annoyingly, there's a bug where certain text is larger in Hires mode than in standard resolution, but that bug is also present without this PR, and I don't think this PR is making it any worse, so I'm not going to block this PR on that.

Nathan-MV commented:
![image](https://github.com/mkxp-z/mkxp-z/assets/67521199/dde63854-3efb-4c3d-94b7-47e21f1a6c02)
![image](https://github.com/mkxp-z/mkxp-z/assets/67521199/76ec54e8-2e99-44b9-abb7-6b5b676f1080)

Nathan-MV commented:
```
1234567890123456789012345678901234567890\C[1]1234567890
\C[0]ABCDEFGHIJKLMNOPQRSTUVWXYZ\C[1]ABCDEFGHIJKLMNOPQRSTUVWXYZ
\C[0]abcdefghijklmnopqrstuvwxyz\C[1]abcdefghijklmnopqrstuvwxyz
```
![image](https://github.com/mkxp-z/mkxp-z/assets/67521199/fee0fa40-2e44-4088-838d-be9b6932e1a0)

Splendide-Imaginarius commented:
@Nathan-MV posting screenshots without any clear explanation of how to reproduce them does not help me determine if this is suitable for merging.

Splendide-Imaginarius commented:
@WaywardHeart Can you rebase this PR now that #98 is merged?

WaywardHeart committed:
Properly handle transparent text with outlines.

RGSS's outline is drawn by blitting a complete set of text four times at different offsets, which affects the final text color and opacity for transparent text. We need to replicate it.

The software blitting this commit introduces actually turned out to be faster than calling TTF_SetFontOutline, which apparently clears the font's glyph cache (which is *not* noted in SDL_ttf.h).

WaywardHeart committed:
Fix position of charLimit parameter in TTF_MeasureUTF8

WaywardHeart commented:
Sorry I took so long to get to this. It was a mix of procrastination and getting distracted by other stuff.

A big recent addition is I finally realized how RGSS is drawing outlines and threw together something to replicate it. Something surprising was that it was actually faster than using TTF_SetFontOutline (I don't break even until the surface's area reaches over 300k pixels). Apparently that function clears the glyph cache, which is _not_ noted anywhere by SDL. So now item menus with a lot of entries should load faster.

Let's see, other changes include config options to control the font scale (Joiplay has it, it seems a little useful) and toggle kerning, tweaks to text positioning (it turns out the outline does affect the position, slightly), and I fixed my usage of TTF_MeasureUTF8. At first I just noticed that I had `charLimit` in the wrong parameter (SDL's description of what they do is in the wrong order, which is probably what threw me off), and after fixing it I noticed some broken characters in games. Apparently it counts the codepoints, not the bytes, so we have to go through and figure out where to cut the string.

I'm also including the `initFromSurface` fix from my new PR, because that really is a major bug.

Eblo commented:
The outline update is a boon. Some comparisons in VX Ace:

Vanilla
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/9230a68f-9369-404d-985d-79ef0bc7e34d)

mkxp-z
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/b13bc1ce-84cc-4893-8427-ae26262c1cd1)

This fix. I found the best settings to be: kerning on, 0.9 scaling, and method 1 (given this is VX Ace, that seems to mean method 2)
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/d4a2b7d1-a03a-40a9-83d2-9308c778fb29)

mkxp-z overlaid on vanilla
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/2dfc5ab7-dc99-41b7-b847-1fc922e6989e)

This fix overlaid on mkxp-z
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/926313c9-4595-4cc2-8806-e7615d3843ce)

This fix overlaid on vanilla
![image](https://github.com/mkxp-z/mkxp-z/assets/7004497/dd6fe7d9-2b2c-4956-aafb-d6f1efdcf645)

The glyphs are still somehow shaped differently from vanilla, seen best with lowercase e. But now that I see this, I'm thinking the outline was the bigger factor in the look and feel of text.

WaywardHeart commented:
> I found the best settings to be: kerning on, 0.9 scaling, and method 1

Oh, right, that being more accurate in your case is from something I noticed and forgot to address: the default font size in VX Ace (I don't know about VX) should be 24, not 22 like XP.

WaywardHeart committed:
The default font size for VX Ace should be 24, not 22

Splendide-Imaginarius commented:
> I'm also including the `initFromSurface` fix from my new PR, because that really is a major bug.

https://github.com/mkxp-z/mkxp-z/pull/191 is merged; can this PR be rebased?

WaywardHeart committed:
Close the RWops if the filesystem fails to open a font.

WaywardHeart commented:
No major code changes since the last update, just making sure a RWops gets closed if the font file disappeared since the initial scan, and making sure we don't try to scale a non-scalable font (which shouldn't be able to happen unless someone renames such a font to .ttf or .otf). I did recently find something juicy, though.

I've looked deeper into how the font size is determined. RPG Maker uses gdi32.dll for its graphics, including drawing text. The part of Wine's implementation that we care about can be found [here](https://github.com/wine-mirror/wine/blob/562aecd9a83f9a34f574a51ca9b24e32a94b303e/dlls/win32u/freetype.c). The `load_VMDK` function in particular is where most of the magic we're looking for happens. A quick look at the output of `Bitmap#text_size` says it usually produces somewhat different widths than windows, but it's still probably the closest we can get.

Wine attempts to load a table in the font called VMDX, which has different metrics for ascent and descent than the font normally provides. If the table isn't found or the requested size is out of range of it, then it computes the proper size with some other formula. The freetype face pointer is stored at the beginning of the `TTF_Font` objects, so I can retrieve the VDMX table purely from our code. Wine also uses `FT_Set_Pixel_Sizes` instead of `FT_Set_Char_Size`, which might produce slightly different results.

Here's a link I found that quickly describes the fallback formula, although it's 21 years old at this point and Wine might have refined it by now: https://lists.gnu.org/archive/html/freetype/2003-02/msg00072.html

I've also realized that RGSS's text squeezing can be replicated by reducing the width. I haven't implemented it yet either, but the knowledge is there, now.

WaywardHeart committed:
Proper font sizing

Windows preferentially uses a special table to find the ppem for a requested height. Code taken from Wine's implementation.

WaywardHeart commented:
I've been sitting on this for a couple weeks and am still feeling too lazy to rebase out the old dpi stuff, so here's this for people to look at. It should be as accurate as I can get it without modifying SDL_ttf.

I poked around a bit at the squeezing since my last update. The Windows api that's used to draw text allows for requesting an average character width, which is computed using the `xAvgCharWidth` property in the font's OS2 table. Based on the behavior of a couple fonts with very different average widths (Arial and [04b_19](https://www.dafont.com/04b-19.font)), RGSS assumes the average char width for each size instead of asking for it (if it even can) or attempting to calculate a good one, which results in wider fonts getting squeezed a lot more on the initial squeeze. Subsequent squeezes also happen very early, which means RGSS is assuming the text width is reduced by a certain percentage.

Splendide-Imaginarius commented:
> RGSS's outline is drawn by blitting a complete set of text four times at different offsets, which affects the final text color and opacity for transparent text. We need to replicate it.

This seems to interact badly with high-res mode, where the outline is typically more than 1 pixel thick. (I briefly tested with a 10x scaling factor, where the outline is 10 pixels thick, and the corners of text look very odd.) Do you have thoughts on this? (I can probably post a screenshot if needed, but I'm guessing you can immediately tell what I mean.)

WaywardHeart commented:
Oh. Yes, I suppose that would look bad. I was more interested in replicating the color of transparent text than perfectly replicating the outline, and doing it this way produces a secondary outline at the edge of the normal text from the outline beneath it being layered twice instead of four times that I remember noticeably affecting the appearance. That probably looks ugly in high-res mode, though, so I guess I'll need to cook up a way of faking it while using SDL_ttf's outline and accept that the edge of transparent text will look a little off compared to RGSS.

Splendide-Imaginarius commented:
@WaywardHeart is there any way we could do what your PR does, but just with more passes with high-res mode than in standard-res mode? E.g. if the outline thickness is 2 pixels, do twice as many outline passes (one at 1px distance, one at 2px distance)? Would that cause a performance problem at 10x?

Splendide-Imaginarius commented:
Converted this PR to a draft until the high-res mode regression is figured out.

@WaywardHeart this PR is getting pretty big, and does a lot of things. Some of which are very clearly uncontroversial and have already passed testing. Would you like to split this PR into multiple PR's so that we can get parts of it merged more promptly?

WaywardHeart committed:
Go back to using SDL_ttf's outlines

Using RGSS's method looks very ugly in hires mode, and there's no easy way to fix it, so instead we need to fake the effect.

To speed up generating large menus, we're now also producing a second TTF_Font object for the outlines.

WaywardHeart committed:
Change how hires text sizing works.

Since the ppem doesn't directly correspond to the size now, it makes more sense to apply the hires multiplier to the ppem instead of the size. This commit's implementation has the side effect of not allowing drawing "lores" text directly to a hires bitmap; whatever size you set will always have the multiplier applied on top of it.

WaywardHeart commented:
I finally got around to working on this. I went back to using SDL_ttf's outlines and faked the transparency effect. ~~I decided to use the average of 2-layers and 4-layers for adjusting the text color - it should be close enough in appearance.~~ I decided averaging like that wasn't close enough and to just not worry about it – it's just a small difference, anyway. I'm also creating a second font object for the outline, to speed up drawing menus with lots of items. We could add even more for bold and italic styles, but the slowdown should only be noticeable if the game constantly switches back and forth like we'd have to for outlines.

Since the font's ppem no longer directly correlates to the size, I'm also changing how hires font sizing works. Now we're storing the multiplier directly in the font object. This has the side effect of meaning that games can't draw "lores" text directly to hires bitmaps anymore, which I'm assuming won't be a problem.

WaywardHeart committed:
I changed my mind about averaging the outline layers for transparent text color. The effect it was supposed to fake was only a small difference, anyway.

WaywardHeart committed:
Use the output of TTF_FontHeight for Bitmap#text_size and text positioning.

WaywardHeart commented:
I consolidated some commits. Font hinting is now disabled when fontSizeMethod is 2, which fixes character spacing differences from Windows. And `Bitmap#text_size` now returns the font height for the height instead of the height SDL_ttf reports, and we also use that height for positioning. I don't entirely like that last one, because it means the bottoms of some characters being cut off, but it matches RGSS's behavior. The vertical alignment is still occasionally off by one, but that problem's still rooted in SDL_ttf and would need to be fixed there (the ascent and descent probably need to be floored instead of ceiled). I should be done fiddling with this, unless someone can find a problem with it. The only thing left to be done is squeezing, and I don't plan on doing that any time soon.

On a minor note, I've realized that fonts with thin stroke widths (like Arial) would only have 2 outline layers behind the text, or even less. Characters that thin are mostly outline, though, so the extra layers shouldn't make much difference, but it's still mildly annoying.

Splendide-Imaginarius commented:
> I'm also creating a second font object for the outline, to speed up drawing menus with lots of items. We could add even more for bold and italic styles, but the slowdown should only be noticeable if the game constantly switches back and forth like we'd have to for outlines.

Could we maybe have a customScript added to the `tests` folder that benchmarks the perf of outlines, bold, and italic?

> I don't entirely like that last one, because it means the bottoms of some characters being cut off, but it matches RGSS's behavior.

IMO bug-for-bug compatibility with Enterbrain's implementation should be optional; could we maybe have a config option for this?

> Some Pokemon Essentials games were explicitly made for mkxp. Text rendered using the old method tends to be smaller than what the new one produces, which means text in those games tends to get cut off with it.
> 
> Since Pokemon Essentials uses RGSS1, the setting defaults to the old method for RGSS1, and the new method for RGSS2 and RGSS3.

While I don't object to bug-for-bug compatibility with Enterbrain's implementation (as long as it can be turned off), I am hesitant to have bug-for-bug compatibility with old mkxp-z versions. It sounds like it would be a precedent for a maintenance nightmare that I don't want to deal with. Is it really that unreasonable to expect game devs who are depending on mkxp-z bugs to either update their code or stick with an old mkxp-z version? (I'll ask about this on Matrix/Discord too; maybe my concern is misplaced.)

Splendide-Imaginarius commented:
> Converted this PR to a draft until the high-res mode regression is figured out.

That seems to be fixed, so un-drafting this PR.

Splendide-Imaginarius commented on src/display/font.cpp:
```diff
@@ -158,14 +194,267 @@ void SharedFontState::initFontSetCB(SDL_RWops &ops,
 		set.other = filename;
 }
 
+/* The following code was taken from Wine to emulate
+ * Windows's font size selection behavior. */
```
A permalink to the Wine code that was used would make this easier to review/maintain.

Splendide-Imaginarius commented:
> While I don't object to bug-for-bug compatibility with Enterbrain's implementation (as long as it can be turned off), I am hesitant to have bug-for-bug compatibility with old mkxp-z versions. It sounds like it would be a precedent for a maintenance nightmare that I don't want to deal with. Is it really that unreasonable to expect game devs who are depending on mkxp-z bugs to either update their code or stick with an old mkxp-z version? (I'll ask about this on Matrix/Discord too; maybe my concern is misplaced.)

General consensus on Matrix/Discord is that bug-for-bug compatibility with Enterbrain's implementation is desirable but should be configurable, but bug-for-bug compatibility with older mkxp-z versions should not be merged due to maintenance concerns.

Splendide-Imaginarius commented:
Superseded by https://github.com/mkxp-z/mkxp-z/pull/283
